### DISCUSSION TITLE
propan run command dont see root packege

### DISCUSSION THREAD

**Discussion by koevgeny10** [Lancetnik/Propan#130] [Questions] [answered]
I have a problem when I place code in packege named "root", separated by modules "main" and "handlers" like this

handlers.py
```python
from propan import RabbitRouter

router = RabbitRouter(prefix="user/")


@router.handle("created")
async def handle_user_created_event(user_id: str):
    ...
```
main.py
```python
from propan import PropanApp, RabbitBroker

from root.handlers import router

broker = RabbitBroker()
broker.include_router(router)
app = PropanApp(broker)
```
Сommand `propan run root.main:app` gives me an error
```
Traceback (most recent call last):

  File "/home/evgeny/projects/education/try/propan/.venv/bin/propan", line 8, in <module>
    sys.exit(cli())
             ^^^^^

  File "/home/evgeny/projects/education/try/propan/.venv/lib/python3.11/site-packages/propan/cli/main.py", line 100, in run
    _run(module=module, app=app, extra_options=extra, log_level=casted_log_level)

  File "/home/evgeny/projects/education/try/propan/.venv/lib/python3.11/site-packages/propan/cli/main.py", line 110, in _run
    propan_app = try_import_propan(module, app)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/home/evgeny/projects/education/try/propan/.venv/lib/python3.11/site-packages/propan/cli/utils/imports.py", line 12, in try_import_propan
    propan_app = import_object(module, app)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/home/evgeny/projects/education/try/propan/.venv/lib/python3.11/site-packages/propan/cli/utils/imports.py", line 40, in import_object
    loader.exec_module(mod)

  File "<frozen importlib._bootstrap_external>", line 940, in exec_module

  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed

  File "/home/evgeny/projects/education/try/propan/root/main.py", line 3, in <module>
    from root.handlers import router

ModuleNotFoundError: No module named 'root'
```
How to split propan code by modules in packege?

**Comment by Lancetnik** [comment]
Propan places your interpreter to the running script root
So the command 'propan run root.main:app' is the same to run 'python main.py' from the 'root' directory
So, your imports should be correct for this case and you can check it the same way

**Reply by koevgeny10** [reply]
I dont understend
```
cd root
propan run main:app
```
gives me the same error

**Reply by Lancetnik** [reply]
Can you show me the director tree?

**Reply by Lancetnik** [reply, marked as answer]
Anyway, you can't import from 'root' if your entry script is inside the 'root'
You need to create subdirectory at the same level with 'main.py' and import all modules from it
Also, you can generate any of Propan project templates to take a look at the project structure

**Reply by koevgeny10** [reply]
> Propan places your interpreter to the running script root

Ok, I seen your example of structure below and structure of template generated by cli and understood. main.py must be on top of src/ and all imports must be relative, like `import from app` but not `import from src.app` (if we imagine that src/ has `__init__.py`). But if we hide main.py in another directory in between then nothing will work. It's strange that library constrains flexibility of python imports baning absolute imports. For example uvicorn has't that constrains.
Am I right?

**Reply by Lancetnik** [reply]
Propan dosn't brake default python interpreter import behavior, it just places import path to the root of the running script
Uvicorn had the same behavior in the past, but now it was moved to app-dir parameter
In the Propan case this feature allows you to run script from any parent directory you want:
```
~/         propan run apps.app.main:app
~/apps     propan run app.main:app
~/apps/app propan run main:app
```

**Comment by Lancetnik** [comment]
As an example: having the following project structure

```
example/
├── .env
├── .gitignore
├── docker-compose.yaml
├── Dockerfile
├── README.md
├── requirements.txt
├── src/
├───── main.py
└───── app/
      ├── apps/
      │   ├── __init__.py
      │   └── handlers.py
      ├── config/
      │   ├── __init__.py
      │   └── settings.py
      └── __init__.py
```

You should use improts from the `app` directory, not `src` one
